src/librbd/cache/ReplicatedWriteLogInternal.h

```diff
@@ -0,0 +1,627 @@
+    op_hist_x_axis_config, op_hist_y_axis_count_config,
+    "Histogram of log append transaction time (nanoseconds) vs. entries appended");
+  plb.add_time_avg(l_librbd_rwl_retire_tx_t, "retire_tx_lat", "Log retire transaction latency");
+  plb.add_u64_counter_histogram(
+    l_librbd_rwl_retire_tx_t_hist, "retire_tx_lat_histogram",
+    op_hist_x_axis_config, op_hist_y_axis_count_config,
+    "Histogram of log retire transaction time (nanoseconds) vs. entries retired");
+
+  m_perfcounter = plb.create_perf_counters();
+  m_image_ctx.cct->get_perfcounters_collection()->add(m_perfcounter);
+}
+
+template <typename I>
+void ReplicatedWriteLog<I>::perf_stop() {
+  assert(m_perfcounter);
+  m_image_ctx.cct->get_perfcounters_collection()->remove(m_perfcounter);
+  delete m_perfcounter;
+}
+
+template <typename I>
+void ReplicatedWriteLog<I>::log_perf() {
+  bufferlist bl;
+  Formatter *f = Formatter::create("json-pretty");
+  bl.append("Perf dump follows\n--- Begin perf dump ---\n");
+  bl.append("{\n");
+  stringstream ss;
+  utime_t now = ceph_clock_now();
+  ss << "\"test_time\": \"" << now << "\",";
+  ss << "\"image\": \"" << m_image_ctx.name << "\",";
+  bl.append(ss);
+  bl.append("\"stats\": ");
+  m_image_ctx.cct->get_perfcounters_collection()->dump_formatted(f, 0);
+  f->flush(bl);
+  bl.append(",\n\"histograms\": ");
+  m_image_ctx.cct->get_perfcounters_collection()->dump_formatted_histograms(f, 0);
+  f->flush(bl);
+  delete f;
+  bl.append("}\n--- End perf dump ---\n");
+  bl.append('\0');
+  ldout(m_image_ctx.cct, 1) << bl.c_str() << dendl;
+}
+
+template <typename I>
+void ReplicatedWriteLog<I>::periodic_stats() {
+  std::lock_guard locker(m_lock);
+  ldout(m_image_ctx.cct, 1) << "STATS: "
+			    << "m_free_log_entries=" << m_free_log_entries << ", "
+			    << "m_log_entries=" << m_log_entries.size() << ", "
+			    << "m_dirty_log_entries=" << m_dirty_log_entries.size() << ", "
+			    << "m_bytes_allocated=" << m_bytes_allocated << ", "
+			    << "m_bytes_cached=" << m_bytes_cached << ", "
+			    << "m_bytes_dirty=" << m_bytes_dirty << ", "
+			    << "bytes available=" << m_bytes_allocated_cap - m_bytes_allocated << ", "
+			    << "m_current_sync_gen=" << m_current_sync_gen << ", "
+			    << "m_flushed_sync_gen=" << m_flushed_sync_gen << ", "
+			    << dendl;
+}
+
+template <typename I>
+void ReplicatedWriteLog<I>::arm_periodic_stats() {
+  assert(m_timer_lock.is_locked());
+  if (m_periodic_stats_enabled) {
+    m_timer.add_event_after(LOG_STATS_INTERVAL_SECONDS, new LambdaContext(
+      [this](int r) {
+	/* m_timer_lock is held */
+	periodic_stats();
+	arm_periodic_stats();
+      }));
+  }
+}
+
+template <typename I>
+void ReplicatedWriteLog<I>::rwl_init(Context *on_finish, DeferredContexts &later) {
+  CephContext *cct = m_image_ctx.cct;
+  ldout(cct, 20) << dendl;
+  TOID(struct WriteLogPoolRoot) pool_root;
+  auto cache_state = dynamic_cast<ImageCacheStateRWL<I>*>(m_cache_state);
```

Comment:
The cache_state is initialized in the constructor, so it mustn't be null.

Reply:
It's a dynamic_cast so it could fail -- if it's always a `ImageCacheStateRWL`, store it as such and avoid the dynamic_cast

Reply:
The m_cache_state inherits from ImageCache. In ImageCache, its type is general *ImageCacheState. I can't find a way to avoid dynamic_cast, do you have any suggestion? @dillaman

Reply:
Why does `ImageCache` need `m_cache_state`?

Reply:
In our design, every image cache which inherits from ImageCache has a cache state. The basic information is cache_type, exists etc and saved in metadata image_cache_state. When the image cache is initialized, it reads out the metadata. And constructs the specified cache class based on cache_type in the metadata.

Reply:
... but you could pull it down into `ReplicatedWriteLog` since nothing else is using it. Then you can avoid the ugly `dynamic_cast`

Reply:
Removed ImageCacheState from ImageCache, and put it into ReplicatedWriteLog class.